stan/math/fwd/fun/hypergeometric_pFq.hpp

```diff
@@ -0,0 +1,219 @@
+ * @param[in] b b
+ * @param[in] z z
+ * @return Generalised hypergeometric function
+ */
+template <typename Ta, typename Tb, typename Tz,
+          require_all_eigen_vector_vt<is_fvar, Ta>* = nullptr,
+          require_all_eigen_vector_vt<std::is_arithmetic, Tb>* = nullptr,
+          require_fvar_t<Tz>* = nullptr>
+inline return_type_t<Ta, Tb, Tz> hypergeometric_pFq(const Ta& a, const Tb& b,
+                                                    const Tz& z) {
+  using fvar_t = return_type_t<Ta, Tb, Tz>;
+  using fvar_inner_t = typename fvar_t::Scalar;
+  promote_scalar_t<fvar_inner_t, Ta> grad_a(a.size());
+  fvar_inner_t grad_z;
+  grad_pFq_pz(grad_a, grad_z, a.val(), b, z.val());
+  return fvar<fvar_inner_t>(hypergeometric_pFq(a.val(), b, z.val()),
+                            dot_product(a.d(), grad_a) + z.d_ * grad_z);
+}
+
+/**
+ * Returns the generalised hypergeometric function applied to the
+ * input arguments.
+ *
+ * Specialization for hypergeometric_pFq(vector_d, vector_d, fvar<T>)
+ *
+ * @tparam Ta Type of a argument
+ * @tparam Tb Type of b argument
+ * @tparam Tz Type of z argument
+ * @param[in] a a
+ * @param[in] b b
+ * @param[in] z z
+ * @return Generalised hypergeometric function
+ */
+template <typename Ta, typename Tb, typename Tz,
+          require_all_eigen_vector_vt<std::is_arithmetic, Ta, Tb>* = nullptr,
+          require_fvar_t<Tz>* = nullptr>
+inline return_type_t<Ta, Tb, Tz> hypergeometric_pFq(const Ta& a, const Tb& b,
+                                                    const Tz& z) {
+  using fvar_t = return_type_t<Ta, Tb, Tz>;
+  using fvar_inner_t = typename fvar_t::Scalar;
+  fvar_inner_t grad_z;
+  grad_pFq_z(grad_z, a, b, z.val());
+  return fvar<fvar_inner_t>(hypergeometric_pFq(a, b, z.val()), z.d_ * grad_z);
+}
+
+/**
+ * Returns the generalised hypergeometric function applied to the
+ * input arguments.
+ *
+ * Specialization for hypergeometric_pFq(vector_d, vector_f, fvar<T>)
+ * where vector_f = Eigen::Matrix<fvar<T>, -1, 1>
+ *
+ * @tparam Ta Type of a argument
+ * @tparam Tb Type of b argument
+ * @tparam Tz Type of z argument
+ * @param[in] a a
+ * @param[in] b b
+ * @param[in] z z
+ * @return Generalised hypergeometric function
+ */
+template <typename Ta, typename Tb, typename Tz,
+          require_all_eigen_vector_vt<std::is_arithmetic, Ta>* = nullptr,
+          require_all_eigen_vector_vt<is_fvar, Tb>* = nullptr,
+          require_fvar_t<Tz>* = nullptr>
+inline return_type_t<Ta, Tb, Tz> hypergeometric_pFq(const Ta& a, const Tb& b,
+                                                    const Tz& z) {
+  using fvar_t = return_type_t<Ta, Tb, Tz>;
+  using fvar_inner_t = typename fvar_t::Scalar;
+  promote_scalar_t<fvar_inner_t, plain_type_t<Tb>> grad_b(b.size());
+  fvar_inner_t grad_z;
+  grad_pFq_qz(grad_b, grad_z, a, b.val(), z.val());
+  return fvar<fvar_inner_t>(hypergeometric_pFq(a, b.val(), z.val()),
+                            dot_product(b.d(), grad_b) + z.d_ * grad_z);
+}
```

Comment:
So the `grad_pFq_qz(grad_b, grad_z, a, b.val(), z.val());` is a little weird to me. Looking at the impl code it seems like you are doing this based on the `grad_*` plain type. Am I reading this right that you only need the gradients for inputs that are autodiffable? If so then I think this would be a good bit cleaner if we had a scheme where `grad_pFq_qz` did it's deduction for the correct function based on whether the inputs where autodiffable. Then it can return a tuple with size equal to the number of autodiffable values passed into it (and doing `value_of()` before going off to the `grad_pFq_qz_impl` function. 

```cpp
template <typename Ta, typename Tb, typename Tz,
          require_all_eigen_vector_vt<std::is_arithmetic, Ta>* = nullptr,
          require_all_eigen_vector_vt<is_fvar, Tb>* = nullptr,
          require_fvar_t<Tz>* = nullptr>
inline return_type_t<Ta, Tb, Tz> hypergeometric_pFq(const Ta& a, const Tb& b,
                                                    const Tz& z) {
  using fvar_t = return_type_t<Ta, Tb, Tz>;
  using fvar_inner_t = typename fvar_t::Scalar;
  // correct grad_pFq_qz chosen based off of which input is autodiffable
  auto grad_tuple = grad_pFq_qz(a, b, z);
  return fvar<fvar_inner_t>(hypergeometric_pFq(a, b.val(), z.val()),
                            dot_product(b.d(), std::get<0>(grad_tuple)) + z.d_ * std::get<1>(grad_tuple));
}
```

Or if it's easier for the impl you can just always return back a tuple of size 3. I think `grad_pFq` could do something like the below where you just detect which ones you need based on whether the values are constant or not like in the lpdf functions.

```cpp
template <typename Ta, typename Tb, typename Tz,
          require_stan_scalar_t<Tz>* = nullptr,
          require_all_eigen_vector_t<Ta, Tb>* = nullptr>
void grad_pFq(const Ta& a, const Tb& b, const Tz& z,
              double precision = 1e-14, int max_steps = 1e6) {
  std::tuple<promote_scalar_t<partials_type_t<Ta>, Ta>,
    promote_scalar_t<partials_type_t<Tb>, Tb>,
    promote_scalar_t<partials_type_t<Tz>, Tz>> ret_tuple;
  internal::grad_pFq_impl<!is_constant<Ta>, !is_constant<Tb>, !is_constant<Tz>>(
      ret_tuple, a, b, z, precision, max_steps);
  return ret_tuple;
}
```

Reply:
^and set the non-used ones to NaN or something like that

Reply:
Oh neat, this is much cleaner. Good call